### ISSUE TITLE
RTask is not using the proxy to download packages

### ISSUE THREAD

**Issue by samthiriot** [openmole/openmole#457]
Dear beloved OpenMole community,

I tried running an RTask requiring an external library in a computer which requires an http proxy to connect the web. 

I copy the test workflow below. 

The execution fails with an error `there is no package called igraph`.  I copy the error below.

I suspect the RTask does not achieves to download the required package because the http proxy passed to OpenMole using the command line option `--proxy` is not transmitted to the task. 
This suscpicion is reinforced by the fact I don't see any option related to http proxy in [the code of the RTask](https://github.com/openmole/openmole/blob/11-dev/openmole/plugins/org.openmole.plugin.task.r/src/main/scala/org/openmole/plugin/task/r/RTask.scala) (but maybe it is passed implicitely in some way): 

Thanks!

Sam

Errors displayed by OpenMole on output
````
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:DO': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:cd': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:do': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:le': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:li#50:co#189:am:xn:xv:LP:sr': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:cs': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:im': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:ke': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:ti': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:se': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:me': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:Co#8:pa#64:AF': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:vb': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:ac': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:po': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:k3': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:k8': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:F3': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:F7': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:FB': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:K2': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:*7': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:%e': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:@7': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 11: unset: `:kd': not a valid identifier
/workdirectory/.tmp/3651fadd-831d-4156-8bc6-80fcd4a2354d/execution2c79ec9c-4faa-4fc4-819a-fd0f71196b98/file1760aa32-c288-4eed-ad54-afcae86d5151.bin/tmp/launch.sh: line 5: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8): No such file or directory
Error in library(igraph) : there is no package called ￢ﾀﾘigraph￢ﾀﾙ
Execution halted
```

Tested script
```
// inputs 
val networkSize = Val[Integer]
val networkNei = Val[Integer]
val networkRewiring = Val[Double]

// R: we run the network 
val averagepathlength = Val[Double]
val diameter = Val[Double]
val networkfile = Val[File]
val nvertices = Val[Double]
val nedges = Val[Double]
val transitivity = Val[Double]

val generateNetworkTask =
  RTask("""
        # define a filename 
        # generate a unique file name
     
        # load the library
        library(igraph)
        
        # generate the network
        g <- igraph::watts.strogatz.game(
                                size=networkSize,nei=networkNei, p=networkRewiring,
                                dim=1,multiple=F,loops=F)
        g <- as.undirected(g)
        g <- simplify(g)
        
        # mesure the properties of the network
        averagepathlength <- average.path.length(g)
        diameter <- as.integer(diameter(g))
        nvertices <- vcount(g)
        nedges <- ecount(g)
        transitivity <- transitivity(g)

        """, 
        libraries = Seq("igraph") 
  ) set (
    
    rInputs += networkSize,
    rInputs += networkNei,
    rInputs += networkRewiring,
    
    rOutputs += averagepathlength,
    rOutputs += diameter,
    rOutputs += nvertices,
    rOutputs += nedges,
    rOutputs += transitivity,

    networkSize := 1000,
    networkNei := 4,
    networkRewiring := 0.15
  )
  
generateNetworkTask hook CSVHook(workDirectory / "measured.csv")
```

**Comment by romainreuillon** [comment]
I think I would go with passing the network service to the container tasks and setting the HTTP_PROXY environment variable in all container. Then it would solve the pb for python, apt....

https://stackoverflow.com/questions/6467277/proxy-setting-for-r#8297685

Do you see any pb with this solution ?

**Comment by romainreuillon** [comment]
bd84b5f8900c717bc76edd05dcf68c918973dd45 should make it work

Could you test it with version 12.0-SNAPSHOT ? I just pushed a docker image.

**Comment by samthiriot** [comment]
Thanks!

I try it right now and keep you posted.

**Comment by samthiriot** [comment]
Hello again,

so I tried it, and did not managed to make it work. 

Yet there is progress, as we now have a clear error stating there is a network issue 👍 
Here is the stacktrace of the error occuring during the "Preparing" phasis:  
```
Network is unreachable (connect failed)
java.net.SocketException: Network is unreachable (connect failed)
	at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399)
	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242)
	at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:403)
	at java.base/java.net.Socket.connect(Socket.java:609)
	at org.apache.http.conn.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:368)
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:142)
	at org.apache.http.impl.conn.BasicHttpClientConnectionManager.connect(BasicHttpClientConnectionManager.java:313)
	at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:393)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:108)
	at container.Registry$HTTP$.execute(Registry.scala:137)
	at container.Registry$Token$.authentication(Registry.scala:177)
	at container.Registry$Token$.withToken(Registry.scala:164)
	at container.Registry$.downloadManifest(Registry.scala:215)
	at container.ImageDownloader$.$anonfun$downloadContainerImage$2(ImageDownloader.scala:121)
	at container.tool.Retry$.retry(Retry.scala:8)
	at container.tool.Retry$.retry(Retry.scala:17)
	at container.ImageDownloader$.downloadContainerImage(ImageDownloader.scala:121)
	at org.openmole.plugin.task.container.ContainerTask$.downloadImage(ContainerTask.scala:77)
	at org.openmole.plugin.task.container.ContainerTask$.localImage(ContainerTask.scala:233)
	at org.openmole.plugin.task.container.ContainerTask$.$anonfun$prepare$1(ContainerTask.scala:213)
	at org.openmole.tool.file.FilePackage$FileDecorator.$anonfun$withLockInDirectory$1(package.scala:412)
	at org.openmole.tool.file.FilePackage$FileDecorator.$anonfun$withLock$5(package.scala:402)
	at org.openmole.tool.stream.package$.withClosable(package.scala:117)
	at org.openmole.tool.file.FilePackage$FileDecorator.$anonfun$withLock$3(package.scala:400)
	at org.openmole.tool.stream.package$.withClosable(package.scala:117)
	at org.openmole.tool.file.FilePackage$FileDecorator.$anonfun$withLock$1(package.scala:399)
	at org.openmole.tool.lock.LockRepository.withLock(LockRepository.scala:54)
	at org.openmole.tool.file.FilePackage$FileDecorator.withLock(package.scala:399)
	at org.openmole.tool.file.FilePackage$FileDecorator.withLockInDirectory(package.scala:412)
	at org.openmole.plugin.task.container.ContainerTask$.prepare(ContainerTask.scala:203)
	at org.openmole.plugin.task.r.RTask$.apply(RTask.scala:81)
	at $line100000019.$read$$iw$$anon$2.run(<console>:31)
	at org.openmole.core.project.Compiled.eval(Project.scala:200)
	at org.openmole.gui.server.core.ApiImpl.$anonfun$synchronousCompilation$7(ApiImpl.scala:355)
	at org.openmole.core.outputmanager.OutputManager$.withStreamOutputs(OutputManager.scala:123)
	at org.openmole.gui.server.core.ApiImpl.$anonfun$synchronousCompilation$6(ApiImpl.scala:355)
	at org.openmole.gui.ext.server.utils$.catchAll(utils.scala:332)
	at org.openmole.gui.server.core.ApiImpl.synchronousCompilation(ApiImpl.scala:355)
	at org.openmole.gui.server.core.ApiImpl.$anonfun$asynchronousCompilation$1(ApiImpl.scala:390)
	at org.openmole.core.threadprovider.ThreadProvider$RunClosure.run(ThreadProvider.scala:24)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

The thing is: how do I specify OpenMole the http proxy? Is there a way for me to transmit the `--proxy` parameter using the docker approach? 

Note the I already defined environment variables HTTP_PROXY and HTTPS_PROXY at the scale of the docker daemon, so it should pass them to docker... (I guess?).

**Comment by romainreuillon** [comment]
Indeed you have to provide the --proxy option to your docker.

```
version: "3"

services:
  openmole:
    image: openmole/openmole:12.0-SNAPSHOT
    command: openmole-docker --proxy $proxy_url
    hostname: openmole
    volumes:
      - ./data/openmole:/var/openmole/
    ports: 
      - "8080:8080"
    privileged: true

```

**Comment by samthiriot** [comment]
Nice, thank Romain.

So I go much far further to another error :-) 
The simulation fails with the stacktrace: 
```
Process exited a non 0 return code (2)
org.openmole.core.exception.UserBadDataError: Process exited a non 0 return code (2)
	at org.openmole.plugin.task.container.ContainerTask$.executeInstall(ContainerTask.scala:225)
	at org.openmole.plugin.task.container.ContainerTask$.$anonfun$prepare$1(ContainerTask.scala:214)
	at org.openmole.tool.file.FilePackage$FileDecorator.$anonfun$withLockInDirectory$1(package.scala:412)
	at org.openmole.tool.file.FilePackage$FileDecorator.$anonfun$withLock$5(package.scala:402)
	at org.openmole.tool.stream.package$.withClosable(package.scala:117)
	at org.openmole.tool.file.FilePackage$FileDecorator.$anonfun$withLock$3(package.scala:400)
	at org.openmole.tool.stream.package$.withClosable(package.scala:117)
	at org.openmole.tool.file.FilePackage$FileDecorator.$anonfun$withLock$1(package.scala:399)
	at org.openmole.tool.lock.LockRepository.withLock(LockRepository.scala:54)
	at org.openmole.tool.file.FilePackage$FileDecorator.withLock(package.scala:399)
	at org.openmole.tool.file.FilePackage$FileDecorator.withLockInDirectory(package.scala:412)
	at org.openmole.plugin.task.container.ContainerTask$.prepare(ContainerTask.scala:203)
	at org.openmole.plugin.task.r.RTask$.apply(RTask.scala:81)
	at $line100000019.$read$$iw$$anon$2.run(<console>:31)
	at org.openmole.core.project.Compiled.eval(Project.scala:200)
	at org.openmole.gui.server.core.ApiImpl.$anonfun$synchronousCompilation$7(ApiImpl.scala:355)
	at org.openmole.core.outputmanager.OutputManager$.withStreamOutputs(OutputManager.scala:123)
	at org.openmole.gui.server.core.ApiImpl.$anonfun$synchronousCompilation$6(ApiImpl.scala:355)
	at org.openmole.gui.ext.server.utils$.catchAll(utils.scala:332)
	at org.openmole.gui.server.core.ApiImpl.synchronousCompilation(ApiImpl.scala:355)
	at org.openmole.gui.server.core.ApiImpl.$anonfun$asynchronousCompilation$1(ApiImpl.scala:390)
	at org.openmole.core.threadprovider.ThreadProvider$RunClosure.run(ThreadProvider.scala:24)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

and more interesting  the output 
```
/_run_commands.sh: 5: export: {e._1}: bad variable name
```

This type of error usually comes from using a different bash interpreted, but I guess that cannot happen with a docker container... ?

**Comment by romainreuillon** [comment]
Great !

Can you provide the OpenMOLE script to reproduce the error ?

**Comment by romainreuillon** [comment]
Using this script it works-on-my-machine® :

```scala
val networkSize = Val[Integer]
val networkNei = Val[Integer]
val networkRewiring = Val[Double]

// R: we run the network 
val averagepathlength = Val[Double]
val diameter = Val[Double]
val networkfile = Val[File]
val nvertices = Val[Double]
val nedges = Val[Double]
val transitivity = Val[Double]

val generateNetworkTask =
  RTask("""
        # define a filename 
        # generate a unique file name
     
        # load the library
        library(igraph)
        
        # generate the network
        g <- igraph::watts.strogatz.game(
                                size=networkSize,nei=networkNei, p=networkRewiring,
                                dim=1,multiple=F,loops=F)
        g <- as.undirected(g)
        g <- simplify(g)
        
        # mesure the properties of the network
        averagepathlength <- average.path.length(g)
        diameter <- as.integer(diameter(g))
        nvertices <- vcount(g)
        nedges <- ecount(g)
        transitivity <- transitivity(g)

        """, 
        libraries = Seq("igraph") 
  ) set (
    
    inputs += (
        networkSize.mapped, networkNei.mapped, networkRewiring.mapped),
    
    outputs += (
        averagepathlength.mapped, diameter.mapped, nvertices.mapped, nedges.mapped, transitivity.mapped),

    networkSize := 1000,
    networkNei := 4,
    networkRewiring := 0.15
  )
  
generateNetworkTask hook CSVHook(workDirectory / "measured.csv")
```

**Comment by romainreuillon** [comment]
Ohh, I think it is linked to a bug in environment variables mapping.... looking into it

**Comment by romainreuillon** [comment]
It should be solved. Could you pull the latest docker openmole/openmole:12.0-SNAPSHOT and try with this version ?

**Comment by samthiriot** [comment]
Thanks, I test it right now and keep you posted

**Comment by samthiriot** [comment]
Hello back,

It's close to work! 


I would suggest to export, not only `http_proxy` and  `HTTP_PROXY`, but also  `https_proxy` and  `HTTPS_PROXY` with the very same values.

**Comment by romainreuillon** [comment]
Hi Sam,

I fixed it in: 2b0330ef6c03c5c06b2ddd4d11140d7ae1bb13e2

I just pushed a new docker image.

Romain

**Comment by samthiriot** [comment]
It works!
It's compiling the R packages now.
Thanks a lot Romain for the reactive support,
Sam
